src/PureScript/Backend/Optimizer/Semantics.purs

```diff
@@ -1034,17 +1038,18 @@ quote = go
     -- Block constructors
     SemLet ident binding k -> do
       let Tuple level ctx' = nextLevel ctx
-      build ctx $ Let ident level (quote (ctx { effect = false }) binding) $ quote ctx' $ k $ NeutLocal ident level
+      build ctx $ Let ident level (quote (ctx { effect = false }) binding) $ quote ctx' $ k $ NeutLocal ident level (Just binding)
     SemLetRec bindings k -> do
       let Tuple level ctx' = nextLevel ctx
-      let neutBindings = (\(Tuple ident _) -> Tuple ident $ defer \_ -> NeutLocal (Just ident) level) <$> bindings
+      -- todo: can we do better than `Nothing` here?
```

Comment:
You would need to change the `Maybe BackendSemantics` in `NeutLocal` to `Maybe (Lazy BackendSemantics)`, where `deref` must force this. Then you could probably tie this knot.

Reply:
There may be a problem here though in the case of a strict recursive loop. A runtime check is emitted with an appropriate exception, but we have no such check as part of eval. Probably best to leave recursive terms out of it for now.